### DISCUSSION TITLE
[Discuss] Flink on Velox supports

### DISCUSSION THREAD

**Discussion by lsyldliu** [facebookincubator/velox#4389] [Q&A]
Hi, guys

I'm focusing on flink sql in flink community, now we are investigate the native vectorized execution solution of batch sql. We found that velox is an universal open vectorized execution plugin, it can support presto and spark now. Moreover, Flink is very similar to Spark, they are both rpc-based and JVM-based compute engine, also the overall architecture is similar, so we think flink on velox is workable. I want to ask whether velox community has interest to support flink? Maybe there are some effort we can do together.

**Comment by pedroerp** [comment]
Hi @lsyldliu, thanks for reaching out! As we work on expanding Velox's reach and growing our open soruce community, we are certainly interested in providing integrations with other widely adopted big data systems such as Flink. 

Some of our close partners from Intel (cc: @davecohn) have expressed interest in the past in having an open source stream processing Velox integration available, so his team might be interested in helping. In addition, the Velox to Spark integration recently created by Intel (called Gluten - https://github.com/oap-project/gluten/) has a similar JVM/JNI integration mode, so maybe some of that work could be reused. 

Let me know how we can help. The community usually communicate via Slack, so if you send me your work email me or @jijufb can send you an invite and we could pick up from there?

**Comment by jijufb** [comment]
Hi @lsyldliu, please send your work email to jiju@fb.com and I can send you invite to access Velox Slack channel.

**Comment by FelixYBW** [comment]
> In addition, the Velox to Spark integration recently created by Intel (called Gluten - [oap-project/gluten](https://github.com/oap-project/gluten/)) has a similar JVM/JNI integration mode, so maybe some of that work could be reused.

@lsyldliu Yes, we definitely can reuse the native part for Spark and Flink, maybe some JVM part as well. You may ping me (@Binwei Yang) in Slack channel and we can have a sync offline.

**Comment by lsyldliu** [comment]
> Hi @lsyldliu, thanks for reaching out! As we work on expanding Velox's reach and growing our open soruce community, we are certainly interested in providing integrations with other widely adopted big data systems such as Flink.
> 
> Some of our close partners from Intel (cc: @davecohn) have expressed interest in the past in having an open source stream processing Velox integration available, so his team might be interested in helping. In addition, the Velox to Spark integration recently created by Intel (called Gluten - https://github.com/oap-project/gluten/) has a similar JVM/JNI integration mode, so maybe some of that work could be reused.
> 
> Let me know how we can help. The community usually communicate via Slack, so if you send me your work email me or @jijufb can send you an invite and we could pick up from there?

Thanks for your attention, I'm very glad to hear that velox community is interested in integration with flink. Due to I'm not familiar with velox currently, so I want to get some help from community about how integrate velox with flink, it seems that Gluten is very good to me. I will study it first.

**Comment by lsyldliu** [comment]
> Hi @lsyldliu, please send your work email to [jiju@fb.com](mailto:jiju@fb.com) and I can send you invite to access Velox Slack channel.

@jijufb Thank you very much, I'm from alibaba, my work email is ron9.liu@gmail.com.

**Comment by lsyldliu** [comment]
> 

@FelixYBW Okay, it sounds like good news to me.

**Comment by Aitozi** [comment]
Hi @lsyldliu @FelixYBW, nice to see the discussion here. Are there any ongoing work around the Velox intergrated with Flink?  We also try to explore the vector execution engine to accelerate the Flink batch/streaming sql

**Comment by FelixYBW** [comment]
Nothing from my side. @lsyldliu  Did you started the work? You may take a look of this PR in substriat: https://github.com/substrait-io/substrait-java/pull/90. It converts spark's logic plan into substrait plan. Currently Gluten used different way to create the substrait plan. We are checking if we can use the same way here.

The biggest issue is spark plan isn't 1:1 mapping to velox plan. In Gluten when we create substrait plan in Gluten, we customized the generated substrait plan to make sure it's 1:1 mapping between substrait plan and Velox's plan.

![image](https://user-images.githubusercontent.com/47296334/196508746-11bbb544-e019-4309-a2de-7995ad2e0ec4.png)

**Comment by qqibrow** [comment]
@FelixYBW This is Lu from Pinterest. We are also exploring whether we can bring velox in flink. could you add me into velox slack? my email is lniu@pinterest.com. Thanks!

**Comment by qqibrow** [comment]
cc @pedroerp

**Reply by pedroerp** [reply]
Hi Lu, I just sent you a Slack invite.

**Reply by qqibrow** [reply]
Thanks!

**Comment by zoudan** [comment]
Hi @pedroerp, I am also interested in it, could you please invite me, my email is [zoudanx@gmail.com](mailto:zoudanx@gmail.com), thanks!

**Reply by pedroerp** [reply]
Invite sent!

**Reply by jnh5y** [reply]
Hi @pedroerp, I'd be interested in track this as well.  My email is jnhuva@gmail.com.  Can I get an invite?

**Reply by ViggoC** [reply]
Me too, @pedroerp could you invite me? viggoc96@gmail.com, thanks.

**Reply by pedroerp** [reply]
Invites sent to you both!

**Reply by wypb** [reply]
Hi @pedroerp, could you please invite me? wyphao.2007@163.com, thanks.

**Reply by pedroerp** [reply]
Sent as well

**Reply by MrMonad** [reply]
can you please invite me as well alt548@g.harvard.edu

**Reply by whutjs** [reply]
@pedroerp Could you please invite me as well, thanks! whutjs@gmail.com

**Reply by pedroerp** [reply]
Invite sent!

**Reply by XiaoqiMa** [reply]
> Hi @pedroerp, could you please invite me? [wyphao.2007@163.com](mailto:wyphao.2007@163.com), thanks.

@wypb  I am also interested in Flink on Velox support, Could you please invite me to the slack? My email is [xiaoqima2013@gmail.com](mailto:xiaoqima2013@gmail.com). Thanks.

**Comment by sharath1709** [comment]
Hi @pedroerp , This is Sharath from UBER and we're also interested in bringing velox to Flink. could you please invite me, my email is dsaisharath@uber.com

**Reply by pedroerp** [reply]
Invite sent.

**Reply by FangYongs** [reply]
Hi @pedroerp , I'm also interested in this issue, please invite me too, email zjureel@gmail.com Thanks

**Reply by lorinlee** [reply]
Hi @pedroerp ,  Could you please invite me as well, thanks! email: lorinlee1996@gmail.com

**Reply by pedroerp** [reply]
Invite sent.

**Comment by waitinfuture** [comment]
Hi @pedroerp , Could you please invite me as well, thanks! waitinfuture@gmail.com

**Reply by pedroerp** [reply]
Invite sent as well.

**Comment by liuyongvs** [comment]
@pedroerp ， Could you please invite me as well, thanks! **liuyongvs2@gmail.com**

**Comment by zjuwangg** [comment]
hi @pedroerp ， Could you please invite me as well, thanks! zjuwangg@gmail.com

**Comment by Poorvankbhatia** [comment]
hi @pedroerp , Could you please send me an invite. Thanks  [puravbhatia9@gmail.com](mailto:puravbhatia9@gmail.com)

**Comment by Myracle** [comment]
Hi, @pedroerp, Could you please invite me to the slack? My email is liujiangangpeng@gmail.com. Thanks.

**Comment by XiaoqiMa** [comment]
Hi, @pedroerp, Could you please invite me to the slack? My email is [xiaoqima2013@gmail.com](mailto:xiaoqima2013@gmail.com). Thanks.

**Comment by pedroerp** [comment]
If you are here for Slack access, please use this Discussion instead:

https://github.com/facebookincubator/velox/discussions/11348

**Comment by FelixYBW** [comment]
The gear has started to turn.

https://github.com/apache/incubator-gluten/pull/8839

**Reply by zhanglistar** [reply]
Yeah, it's a good start, still more details needs to be cleared of state, stateful functions, retract stream.

**Comment by zhanglistar** [comment]
```
Because stream processing aggregations require engine-specific
logic related to temporal windows, namely tumbling, hopping and
session windows, aggregations are implemented as an extension to
Velox in XStream. Even though these special aggregations are most
commonly required in stream processing applications, there are
plans to provide this operation as part of the core Velox library and
expose it to Presto and Spark users as temporal extensions. This
work is only possible without substantial duplication of effort due
to the unified execution engine provided by Velox.
```
Hi @pedroerp As velox paper says, do you have plan to opensource these stateful functions? Thanks in adance.

**Reply by liuyongvs** [reply]
@pedroerp looks forward to your response, thank you very much!

**Reply by pedroerp** [reply]
@liuyongvs we have some of this logic as part of window functions in Velox already, but no plans to open source the other special aggregation extensions yet, as they are still quite tangled with our internal stream processing system. 

@xiaoxmeng has recently added support for pipeline barriers though, which enables your to implement checkpoints and barriers.
